### PR TITLE
Fixed required attribute not being correctly recognized by form dialog

### LINKED ISSUE #586: Boolean Attributes (like required) not correctly recognized by form dialogs
## Are you reporting a feature or a bug?

Bug

## [Check if the issue is already reported](http://docs.ckeditor.com/#!/guide/dev_issues_readme-section-avoid-duplicates)

http://dev.ckeditor.com/ticket/14291

## Provide detailed reproduction steps (if any)

1. Go to http://sdk.ckeditor.com/samples/fullpreset.html.
1. Open Source mode and replace content with `<input checked required type="checkbox" />`.
1. Switch back to WYSIWYG mode.
1. Double click on checkbox to open dialog.

### Expected result

Dialog should show that the input is required (**required** option checked).

### Actual result

The **required** option is not checked.

## Other details

* Browser: All
* OS: All
* CKEditor version: 4.7.1
* Installed CKEditor plugins: Full preset

The cause of the issue is described here http://dev.ckeditor.com/ticket/14291#comment:2.

### PR DESCRIPTION
**Pull request by engineering-this**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

I've changed form dialog to set its property `required = true` whenever matching attribute exists on form element.

Closes #586

### CODE REVIEW

**Comment by Comandeer** [comment]
Please also rebase onto latest `major`.

**Comment by Comandeer** [comment]
Changed base to `master` and rebased.

File: plugins/forms/plugin.js [hunk 1]
```diff
@@ -277,3 +277,16 @@ CKEDITOR.plugins.add( 'forms', {
...
 } );
+
+CKEDITOR.plugins.forms = {
+	/*
+	* Private helper function for setting dialog required value.
+	*
+	* @private
+	* @param {CKEDITOR.dom.element} element Element which required attribute is checked.
+	* */
+	_setupFormDialogRequired: function( element ) {
```
Comment (Comandeer) [inline plugins/forms/plugin.js]: This is incorrect implementation. According to the [HTML LS standard](https://html.spec.whatwg.org/multipage/input.html#the-required-attribute):

> The required attribute is a boolean attribute. When specified, the element is required.

[And further](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute):

> A number of attributes are boolean attributes. The presence of a boolean attribute on an element represents the true value, and the absence of the attribute represents the false value.

Therefore the correct test is to check if the element has appropriate attribute (`element.hasAttribute( 'required' )`), as even incorrect values are treated as true.

File: plugins/forms/plugin.js [hunk 2]
```diff
@@ -277,3 +277,16 @@ CKEDITOR.plugins.add( 'forms', {
 		}
 	}
 } );
+
+CKEDITOR.plugins.forms = {
```
Comment (Comandeer) [inline plugins/forms/plugin.js]: If we add new namespace, we should also add API docs for it.

File: tests/plugins/forms/manual/required.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+@bender-tags: editor, bug, 4.10.0, forms, 586
```
Comment (Comandeer) [inline tests/plugins/forms/manual/required.md]: Please update version tag and remove redundant `forms` one.

**Comment by engineering-this** [comment]
- Changed base to next.
- Changed check to `hasAttribute`, so either pressence or any value of 'required' attribute on form element will be reflected in dialog.
- Corrected docs entry.
- Refactored tests by extracting duplicate code to global helper.

File: plugins/forms/plugin.js [hunk 4]
```diff
@@ -277,3 +277,24 @@ CKEDITOR.plugins.add( 'forms', {
...
+ */
+CKEDITOR.plugins.forms = {
+	/**
+	* Sets dialog required value.
+	*
+	* @since 4.10.2
+	* @private
+	* @param {CKEDITOR.dom.element} element An element which required attribute is checked.
+	*/
+	_setupFormDialogRequired: function( element ) {
```
Comment (Comandeer) [inline plugins/forms/plugin.js]: That name rubs me up the wrong way… Maybe something like `setupRequiredAttribute`?

File: plugins/forms/plugin.js [hunk 5]
```diff
@@ -277,3 +277,24 @@ CKEDITOR.plugins.add( 'forms', {
...
+ * Namespace containing helper functions for Forms Plugin.
+ *
+ * @since 4.10.2
+ * @private
+ * @singleton
+ * @class CKEDITOR.plugins.forms
+ */
+CKEDITOR.plugins.forms = {
+	/**
+	* Sets dialog required value.
```
Comment (Comandeer) [inline plugins/forms/plugin.js]: Maybe add some more info, e.g. it's based on algorithm described in specification.

File: tests/plugins/forms/manual/required.html [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<body>
+<div id="editor">
+	<p>
+		<input required type="checkbox">
```
Comment (Comandeer) [inline tests/plugins/forms/manual/required.html]: I'd be nice to have more than one `[required]` form, e.g. `[required=whatever]`, `[required=false]` etc.

File: CHANGES.md [hunk 7]
```diff
@@ -6,6 +6,7 @@
 Fixed Issues:
 
 * [#1181](https://github.com/ckeditor/ckeditor-dev/issues/1181): [Chrome] Fixed: Opening context menu in readonly editor results in error.
+* [#586](https://github.com/ckeditor/ckeditor-dev/issues/586) Fixed: required attribute not being correctly recognized by form dialog.
```
Comment (Comandeer) [inline CHANGES.md]: Please add reference to the original contributor.

File: tests/plugins/forms/_helpers/tools.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+( function() {
+	'use strict';
+	window.formsTools = {
+		/**
+		 * Asserts 'required' attribute of a form element.
+		 *
+		 * @param {Object} setup Test setup object.
```
Comment (Comandeer) [inline tests/plugins/forms/_helpers/tools.js]: Why not simply `options`? We use this word everywhere else.

**Comment by engineering-this** [comment]
Updated accordingly to requested changes, no changes to way anything works.

**Pull request merged**